lnrpc/routerrpc/router_server.go

```diff
@@ -262,3 +273,169 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
 		TimeLockDelay:  int64(route.TotalTimeLock),
 	}, nil
 }
+
+// SendToRoute sends a payment through a predefined route. The response of this
+// call contains structured error information.
+func (s *Server) SendToRoute(ctx context.Context,
+	req *SendToRouteRequest) (*SendToRouteResponse, error) {
+
+	if req.Route == nil {
+		return nil, fmt.Errorf("unable to send, no routes provided")
+	}
+
+	route, err := s.cfg.RouterBackend.UnmarshallRoute(req.Route)
+	if err != nil {
+		return nil, err
+	}
+
+	hash, err := lntypes.MakeHash(req.PaymentHash)
+	if err != nil {
+		return nil, err
+	}
+
+	preimage, err := s.cfg.Router.SendToRoute(hash, route)
+
+	// In the success case, return the preimage.
+	if err == nil {
+		return &SendToRouteResponse{
+			Preimage: preimage[:],
+		}, nil
+	}
+
+	// In the failure case, marshall the failure message to the rpc format
+	// before returning it to the caller.
+	rpcErr, err := marshallError(err)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SendToRouteResponse{
+		Failure: rpcErr,
+	}, nil
+}
+
+// marshallError marshall an error as received from the switch to rpc structs
+// suitable for returning to the caller of an rpc method.
+//
+// Because of difficulties with using protobuf oneof constructs in some
+// languages, the decision was made here to use a single message format for all
```

Comment:
what languages is this a problem?

Reply:
In golang it isn't very intuitive.

Reply:
In the PR description there is also some explanation of this.

Reply:
The latest version actually makes it a lot easier, they added new accessor methods, so you just need to check if their nil or not: https://github.com/lightningnetwork/lnd/blob/f8c824fb1d6c5ef8524148f59ea5650af65af98b/rpcserver.go#L4942

Reply:
That is better indeed. But I don't know how it is in other languages though. And there are still the concerns about json serialization if we decide to add a rest api to it.

My preference is to keep it simple. There are more data structures in the rpc interface that contain fields that may be unused.

However if it is a deal breaker for someone who is sure we will be better off with the `oneof` types, let me know and I will convert it.

Reply:
Even within a particular failure message, there are different variations depending on the node software version. Like in the case of the `invalid_or_unknown_payment`.